src/core/qgsnetworkcontentfetcherregistry.cpp

```diff
@@ -0,0 +1,191 @@
+/***************************************************************************
+                       qgsnetworkcontentfetcherregistry.cpp
+                             -------------------
+    begin                : April, 2018
+    copyright            : (C) 2018 by Denis Rouzaud
+    email                : denis@opengis.ch
+
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsnetworkcontentfetcherregistry.h"
+
+#include "qgsapplication.h"
+
+QgsNetworkContentFetcherRegistry::QgsNetworkContentFetcherRegistry()
+  : QObject()
+{
+}
+
+QgsNetworkContentFetcherRegistry::~QgsNetworkContentFetcherRegistry()
+{
+  QMap<QUrl, QgsFetchedContent *>::const_iterator it = mFileRegistry.constBegin();
+  for ( ; it != mFileRegistry.constEnd(); ++it )
+  {
+    it.value()->mFile->close();
+    delete it.value()->mFile;
+  }
+  mFileRegistry.clear();
+}
+
+const QgsFetchedContent *QgsNetworkContentFetcherRegistry::fetch( const QUrl &url, const FetchingMode fetchingMode )
+{
+  QMutexLocker locker( &mMutex );
+  if ( mFileRegistry.contains( url ) )
+  {
+    return mFileRegistry.value( url );
+  }
+
+  QgsFetchedContent *content = new QgsFetchedContent( nullptr, QgsFetchedContent::NotStarted );
+
+  // start
+  QObject::connect( content, &QgsFetchedContent::downloadStarted, this, [ = ]( const bool redownload )
+  {
+    if ( redownload && content->status() == QgsFetchedContent::Downloading )
+    {
+      {
+        QMutexLocker locker( &mMutex );
+        if ( content->mFetchingTask )
+          disconnect( content->mFetchingTask, &QgsNetworkContentFetcherTask::fetched, content, &QgsFetchedContent::taskCompleted );
+      }
+      // no locker when calling cancel!
+      content->cancel();
```

Comment:
I think there's an issue here - you call cancel outside the lock here - so in the time between this and the cancel lambda execution below, the task may be completed and deleted... ( i think?)

Reply:
yes, but I don't see any issue here. The cancel method called here is on the content and not on the task.
When going into `cancelTriggered` slot, it's checked that the task is still alive.

Reply:
Ok - I think we're all good here. Tasks are deleteLatered() anyway, so all the slots should have already been handled.